File: apps/api/services/provisioning_catalog.py
from dataclasses import dataclass

from core.platform_apps import BILLING_APP_SLUG
from domains.provisioning.schemas import (
    ProvisioningDraftReplace,
    ProvisioningFieldDefinition,
    ProvisioningResourceDefinition,
    ProvisioningTargetType,
    ProvisioningValidationIssue,
    ProvisioningValueType,
)


@dataclass(frozen=True)
class _Field:
    label: str
    value_type: ProvisioningValueType
    required: bool = True
    reference_namespace: str | None = None
    allowed_values: tuple[str, ...] | None = None
    unique: bool = False


@dataclass(frozen=True)
class _Resource:
    label: str
    description: str
    multiple: bool
    minimum_items: int
    fields: dict[str, _Field]


FINANCE_RESOURCES: dict[str, _Resource] = {
    "currency": _Resource(
        label="Currencies",
        description="Currencies created for every new finance workspace.",
        multiple=True,
        minimum_items=1,
        fields={
            "code": _Field("ISO code", "string"),
            "name": _Field("Name", "string"),
            "numericCode": _Field("Numeric code", "string", required=False),
            "minorUnit": _Field("Minor unit", "integer"),
            "symbol": _Field("Symbol", "string", required=False),
        },
    ),
    "workspace": _Resource(
        label="Workspace defaults",
        description="Locale and currency defaults for a finance workspace.",
        multiple=False,
        minimum_items=1,
        fields={
            "countryCode": _Field("Country", "reference", reference_namespace="country"),
            "baseCurrency": _Field("Base currency", "reference", reference_namespace="currency"),
            "defaultCurrency": _Field("Default currency", "reference", reference_namespace="currency"),
            "defaultLanguage": _Field("Default language", "reference", reference_namespace="language"),
        },
    ),
    "payment_mode": _Resource(
        label="Payment modes",
        description="Payment methods available on new workspaces.",
        multiple=True,
        minimum_items=1,
        fields={"name": _Field("Name", "string")},
    ),
    "payment_term": _Resource(
        label="Payment terms",
        description="Reusable due-date rules for invoices and estimates.",
        multiple=True,
        minimum_items=1,
        fields={
            "name": _Field("Name", "string"),
            "rule": _Field(
                "Rule",
                "string",
                allowed_values=("DUE_ON_RECEIPT", "NET_DAYS", "END_OF_MONTH", "END_OF_NEXT_MONTH"),
            ),
            "dueDays": _Field("Due days", "integer"),
        },
    ),
    "invoice_preference": _Resource(
        label="Invoice preferences",
        description="Default invoice behavior; tenant edits remain authoritative.",
        multiple=False,
        minimum_items=1,
        fields={
            "defaultTaxBehavior": _Field("Tax behavior", "string", allowed_values=("EXCLUSIVE", "INCLUSIVE")),
            "lateFeesEnabled": _Field("Late fees enabled", "boolean"),
            "lateFeeCalculationType": _Field("Late fee calculation", "string", allowed_values=("PERCENTAGE", "FIXED")),
            "lateFeePercent": _Field("Late fee percent", "decimal", required=False),
            "lateFeeAmount": _Field("Late fee amount", "integer", required=False),
            "lateFeeGraceDays": _Field("Grace days", "integer"),
            "lateFeeGenerateAsDraft": _Field("Generate as draft", "boolean"),
        },
    ),
    "tax_authority": _Resource(
        label="Tax authorities",
        description="Tax administrations available to newly created organizations.",
        multiple=True,
        minimum_items=1,
        fields={
            "name": _Field("Name", "string"),
            "description": _Field("Description", "string", required=False),
            "countryCode": _Field("Country", "reference", reference_namespace="country"),
        },
    ),
    "tax_rate": _Resource(
        label="Tax rates",
        description="Tax rates created for newly provisioned finance workspaces.",
        multiple=True,
        minimum_items=1,
        fields={
            "name": _Field("Name", "string"),
            "description": _Field("Description", "string", required=False),
            "taxType": _Field("Tax type", "string", required=False),
            "rate": _Field("Rate", "decimal"),
            "inclusive": _Field("Inclusive", "boolean"),
            "authority": _Field(
                "Tax authority",
                "reference",
                reference_namespace="tax_authority",
            ),
        },
    ),
}

ORGANIZATION_RESOURCES: dict[str, _Resource] = {
    "organization_profile": _Resource(
        label="Organization profile",
        description="Defaults for the global organization record.",
        multiple=False,
        minimum_items=0,
        fields={
            "countryCode": _Field("Country", "reference", reference_namespace="country"),
            "language": _Field("Language", "string"),
            "timezone": _Field("Timezone", "string"),
        },
    )
}

APPLICATION_RESOURCES: dict[str, dict[str, _Resource]] = {
    BILLING_APP_SLUG: {
        "document_preference": _Resource(
            label="Document preferences",
            description=(
                "Default customer note and terms per document type. Subscription-generated "
                "invoices inherit the invoice preference."
            ),
            multiple=True,
            minimum_items=0,
            fields={
                "documentType": _Field(
                    "Document type",
                    "string",
                    allowed_values=("invoice", "quote", "estimate", "credit_note"),
                    unique=True,
                ),
                "customerNote": _Field("Customer note", "string", required=False),
                "termsAndConditions": _Field("Terms and conditions", "string", required=False),
            },
        )
    }
}


def resource_registry(target_type: ProvisioningTargetType, target_key: str) -> dict[str, _Resource]:
    if target_type == "finance":
        return FINANCE_RESOURCES
    if target_type == "organization":
        return ORGANIZATION_RESOURCES
    return APPLICATION_RESOURCES.get(target_key, {})


def catalog_definitions(
    target_type: ProvisioningTargetType,
    target_key: str,
) -> list[ProvisioningResourceDefinition]:
    definitions: list[ProvisioningResourceDefinition] = []
    for resource_type, resource in resource_registry(target_type, target_key).items():
        definitions.append(
            ProvisioningResourceDefinition(
                resource_type=resource_type,
                label=resource.label,
                description=resource.description,
                multiple=resource.multiple,
                minimum_items=resource.minimum_items,
                maximum_items=None if resource.multiple else 1,
                fields=[
                    ProvisioningFieldDefinition(
                        key=key,
                        label=field.label,
                        value_type=field.value_type,
                        required=field.required,
                        reference_namespace=field.reference_namespace,
                        allowed_values=list(field.allowed_values) if field.allowed_values else None,
                    )
                    for key, field in resource.fields.items()
                ],
            )
        )
    return definitions


def validate_draft(
    target_type: ProvisioningTargetType,
    target_key: str,
    draft: ProvisioningDraftReplace,
) -> list[ProvisioningValidationIssue]:
    issues: list[ProvisioningValidationIssue] = []
    if target_type != "application" and draft.finance_dependency != "none":
        issues.append(
            ProvisioningValidationIssue(
                path="finance_dependency",
                code="invalid_target_contract",
                message="Only application manifests may declare an embedded finance dependency.",
            )
        )

    registry = resource_registry(target_type, target_key)
    counts: dict[str, int] = {}
    for index, resource in enumerate(draft.resources):
        path = f"resources.{index}"
        definition = registry.get(resource.resource_type)
        if definition is None:
            issues.append(
                ProvisioningValidationIssue(
                    path=f"{path}.resource_type",
                    code="unknown_resource_type",
                    message=f"Resource type '{resource.resource_type}' is not registered for this target.",
                )
            )
            continue

        counts[resource.resource_type] = counts.get(resource.resource_type, 0) + 1
        values = {property_.key: property_ for property_ in resource.properties}
        for key, field in definition.fields.items():
            property_ = values.get(key)
            if property_ is None:
                if field.required:
                    issues.append(
                        ProvisioningValidationIssue(
                            path=f"{path}.properties.{key}",
                            code="missing_property",
                            message=f"Property '{key}' is required.",
                        )
                    )
                continue
            if property_.value_type != field.value_type:
                issues.append(
                    ProvisioningValidationIssue(
                        path=f"{path}.properties.{key}.value_type",
                        code="invalid_property_type",
                        message=f"Property '{key}' must use value type '{field.value_type}'.",
                    )
                )
            if field.reference_namespace and property_.reference_namespace != field.reference_namespace:
                issues.append(
                    ProvisioningValidationIssue(
                        path=f"{path}.properties.{key}.reference_namespace",
                        code="invalid_reference_namespace",
                        message=f"Property '{key}' must reference '{field.reference_namespace}'.",
                    )
                )
            scalar = property_.string_value
            if field.allowed_values and scalar not in field.allowed_values:
                issues.append(
                    ProvisioningValidationIssue(
                        path=f"{path}.properties.{key}",
                        code="invalid_property_value",
                        message=f"Property '{key}' must be one of: {', '.join(field.allowed_values)}.",
                    )
                )
        for key in values.keys() - definition.fields.keys():
            issues.append(
                ProvisioningValidationIssue(
                    path=f"{path}.properties.{key}",
                    code="unknown_property",
                    message=f"Property '{key}' is not registered for resource type '{resource.resource_type}'.",
                )
            )

    for resource_type, count in counts.items():
        if count > 1 and not registry[resource_type].multiple:
            issues.append(
                ProvisioningValidationIssue(
                    path="resources",
                    code="resource_cardinality",
                    message=f"Resource type '{resource_type}' permits only one row.",
                )
            )
    for resource_type, definition in registry.items():
        count = counts.get(resource_type, 0)
        if count < definition.minimum_items:
            issues.append(
                ProvisioningValidationIssue(
                    path="resources",
                    code="resource_minimum",
                    message=f"Resource type '{resource_type}' requires at least {definition.minimum_items} row(s).",
                )
            )

    reference_namespaces = {
        field.reference_namespace
        for definition in registry.values()
        for field in definition.fields.values()
        if field.reference_namespace in registry
    }
    reference_targets = {
        resource_type: {resource.key for resource in draft.resources if resource.resource_type == resource_type}
        for resource_type in reference_namespaces
    }
    for index, resource in enumerate(draft.resources):
        for property_ in resource.properties:
            namespace = property_.reference_namespace
            if namespace in reference_targets and property_.reference_key not in reference_targets[namespace]:
                issues.append(
                    ProvisioningValidationIssue(
                        path=f"resources.{index}.properties.{property_.key}",
                        code="unresolved_reference",
                        message=f"Reference '{namespace}/{property_.reference_key}' does not exist in this draft.",
                    )
                )

    for resource_type, definition in registry.items():
        for field_key, field in definition.fields.items():
            if not field.unique:
                continue
            seen: set[tuple[object, ...]] = set()
            for index, resource in enumerate(draft.resources):
                if resource.resource_type != resource_type:
                    continue
                property_ = next(
                    (candidate for candidate in resource.properties if candidate.key == field_key),
                    None,
                )
                if property_ is None:
                    continue
                value = (
                    property_.value_type,
                    property_.string_value,
                    property_.integer_value,
                    property_.decimal_value,
                    property_.boolean_value,
                    property_.reference_namespace,
                    property_.reference_key,
                )
                if value in seen:
                    issues.append(
                        ProvisioningValidationIssue(
                            path=f"resources.{index}.properties.{field_key}",
                            code="duplicate_unique_property",
                            message=(
                                f"Property '{field_key}' must be unique within resource type '{resource_type}'."
                            ),
                        )
                    )
                else:
                    seen.add(value)
    return issues

File: apps/api/src/services/provisioning-catalog.ts
import type { ProvisioningDraftReplaceParams as ProvisioningWireDraftReplace } from '@876/core/types/provisioning'

export const BILLING_APP_SLUG = '876-billing'
const CRM_APP_SLUG = '876-crm'
const ISO_DATE_PATTERN = /^\d{4}-\d{2}-\d{2}$/

// Provisioning catalog — the closed set of resource types each provisioning
// target may declare, and the validation that keeps a manifest-v1 draft
// consistent. The catalog describes allowed configuration; operational values
// live in database-backed provisioning manifests.

export type ProvisioningTargetType = 'organization' | 'finance' | 'application'
export type ProvisioningValueType =
  'string' | 'integer' | 'decimal' | 'boolean' | 'reference'
export type FinanceDependency = 'none' | 'embedded'

export type ProvisioningFieldDefinition = {
  key: string
  label: string
  valueType: ProvisioningValueType
  required: boolean
  referenceNamespace: string | null
  allowedValues: string[] | null
}

export type ProvisioningResourceDefinition = {
  resourceType: string
  label: string
  description: string
  multiple: boolean
  minimumItems: number
  maximumItems: number | null
  fields: ProvisioningFieldDefinition[]
}

export type ProvisioningValidationIssue = {
  path: string
  code: string
  message: string
}

export type ProvisioningPropertyInput = {
  key: string
  valueType: ProvisioningValueType
  stringValue: string | null
  integerValue: number | null
  decimalValue: string | null
  booleanValue: boolean | null
  referenceNamespace: string | null
  referenceKey: string | null
}

export type ProvisioningResourceInput = {
  resourceType: string
  key: string
  position: number
  properties: ProvisioningPropertyInput[]
}

export type ProvisioningDraftReplace = {
  manifestVersion?: 1
  reconciliation?: 'create_missing'
  preserveTenantOverrides?: true
  financeDependency: FinanceDependency
  financeScopes?: string[]
  resources: ProvisioningResourceInput[]
  steps?: unknown[]
}

/**
 * Convert the public manifest-v1 wire contract into the catalog's internal
 * camel-case representation. API schemas, bounded clients and handoff imports
 * all use the public snake-case contract; catalog validation must never rely on
 * casts between those shapes.
 */
export function provisioningDraftForCatalog(
  draft: ProvisioningWireDraftReplace
): ProvisioningDraftReplace {
  return {
    manifestVersion: draft.manifest_version ?? 1,
    reconciliation: draft.reconciliation ?? 'create_missing',
    preserveTenantOverrides: draft.preserve_tenant_overrides ?? true,
    financeDependency: draft.finance_dependency ?? 'none',
    financeScopes: draft.finance_scopes ?? [],
    resources: (draft.resources ?? []).map((resource) => ({
      resourceType: resource.resource_type,
      key: resource.key,
      position: resource.position,
      properties: resource.properties.map((property) => ({
        key: property.key,
        valueType: property.value_type,
        stringValue: property.string_value ?? null,
        integerValue:
          property.integer_value === null ||
          property.integer_value === undefined
            ? null
            : Number(property.integer_value),
        decimalValue: property.decimal_value ?? null,
        booleanValue: property.boolean_value ?? null,
        referenceNamespace: property.reference_namespace ?? null,
        referenceKey: property.reference_key ?? null,
      })),
    })),
    steps: draft.steps ?? [],
  }
}

type Field = {
  label: string
  valueType: ProvisioningValueType
  required: boolean
  referenceNamespace: string | null
  allowedValues: readonly string[] | null
  unique: boolean
}

type Resource = {
  label: string
  description: string
  multiple: boolean
  minimumItems: number
  fields: Record<string, Field>
}

function field(
  label: string,
  valueType: ProvisioningValueType,
  opts: Partial<Omit<Field, 'label' | 'valueType'>> = {}
): Field {
  return {
    label,
    valueType,
    required: opts.required ?? true,
    referenceNamespace: opts.referenceNamespace ?? null,
    allowedValues: opts.allowedValues ?? null,
    unique: opts.unique ?? false,
  }
}

function resource(
  label: string,
  description: string,
  multiple: boolean,
  minimumItems: number,
  fields: Record<string, Field>
): Resource {
  return { label, description, multiple, minimumItems, fields }
}

export const FINANCE_RESOURCES: Record<string, Resource> = {
  workspace: resource(
    'Workspace',
    'Locale and currency defaults for a finance workspace.',
    false,
    1,
    {
      countryCode: field('Country', 'reference', {
        required: false,
        referenceNamespace: 'country',
      }),
      baseCurrency: field('Base currency', 'reference', {
        referenceNamespace: 'currency',
      }),
      defaultCurrency: field('Default currency', 'reference', {
        referenceNamespace: 'currency',
      }),
      defaultLanguage: field('Default language', 'reference', {
        referenceNamespace: 'language',
      }),
    }
  ),
  currency: resource(
    'Currencies',
    'Currencies created for every new finance workspace.',
    true,
    1,
    {
      code: field('ISO code', 'string'),
      name: field('Name', 'string'),
      numericCode: field('Numeric code', 'string', { required: false }),
      minorUnit: field('Minor unit', 'integer'),
      symbol: field('Symbol', 'string', { required: false }),
    }
  ),
  payment_mode: resource(
    'Payment modes',
    'Payment methods available on new workspaces.',
    true,
    1,
    { name: field('Name', 'string') }
  ),
  payment_term: resource(
    'Payment terms',
    'Reusable due-date rules for invoices and estimates.',
    true,
    1,
    {
      name: field('Name', 'string'),
      rule: field('Rule', 'string', {
        allowedValues: [
          'DUE_ON_RECEIPT',
          'NET_DAYS',
          'END_OF_MONTH',
          'END_OF_NEXT_MONTH',
        ],
      }),
      dueDays: field('Due days', 'integer'),
    }
  ),
  invoice_preference: resource(
    'Invoice preferences',
    'Default invoice behavior; tenant edits remain authoritative.',
    false,
    1,
    {
      defaultTaxBehavior: field('Tax behavior', 'string', {
        allowedValues: ['EXCLUSIVE', 'INCLUSIVE'],
      }),
      lateFeesEnabled: field('Late fees enabled', 'boolean'),
      lateFeeCalculationType: field('Late fee calculation', 'string', {
        allowedValues: ['PERCENTAGE', 'FIXED'],
      }),
      lateFeePercent: field('Late fee percent', 'decimal', { required: false }),
      lateFeeAmount: field('Late fee amount', 'integer', { required: false }),
      lateFeeGraceDays: field('Grace days', 'integer'),
      lateFeeGenerateAsDraft: field('Generate as draft', 'boolean'),
    }
  ),
  tax_code: resource(
    'Tax applicability',
    'Optional tax categories/scopes used to associate rates with goods or services; this is configuration metadata, not a calculation engine.',
    true,
    0,
    {
      name: field('Name', 'string'),
      description: field('Description', 'string', { required: false }),
      code: field('Code', 'string', { required: false, unique: true }),
      scope: field('Scope', 'string', { required: false }),
    }
  ),
  tax_authority: resource(
    'Tax authorities',
    'Tax administrations available to newly created organizations.',
    true,
    0,
    {
      name: field('Name', 'string'),
      description: field('Description', 'string', { required: false }),
      countryCode: field('Country', 'reference', {
        referenceNamespace: 'country',
      }),
      jurisdiction: field('Jurisdiction', 'reference', {
        required: false,
        referenceNamespace: 'tax_jurisdiction',
      }),
    }
  ),
  tax_jurisdiction: resource(
    'Tax jurisdictions',
    'Geographic tax areas used to model national and sub-national tax rules.',
    true,
    0,
    {
      name: field('Name', 'string'),
      countryCode: field('Country', 'reference', {
        referenceNamespace: 'country',
      }),
      type: field('Jurisdiction type', 'string', {
        allowedValues: [
          'country',
          'state',
          'province',
          'county',
          'city',
          'district',
          'other',
        ],
      }),
      code: field('Jurisdiction code', 'string', { required: false }),
      parent: field('Parent jurisdiction', 'reference', {
        required: false,
        referenceNamespace: 'tax_jurisdiction',
      }),
    }
  ),
  tax_rate: resource(
    'Tax rates',
    'Tax rates created for newly provisioned finance workspaces.',
    true,
    0,
    {
      name: field('Name', 'string'),
      description: field('Description', 'string', { required: false }),
      taxType: field('Tax type', 'string', { required: false }),
      rate: field('Rate', 'decimal'),
      inclusive: field('Inclusive', 'boolean'),
      authority: field('Tax authority', 'reference', {
        required: false,
        referenceNamespace: 'tax_authority',
      }),
      jurisdiction: field('Jurisdiction', 'reference', {
        required: false,
        referenceNamespace: 'tax_jurisdiction',
      }),
      taxCode: field('Tax applicability', 'reference', {
        required: false,
        referenceNamespace: 'tax_code',
      }),
      effectiveFrom: field('Effective from', 'string', { required: false }),
      effectiveUntil: field('Effective until', 'string', { required: false }),
    }
  ),
}

export const ORGANIZATION_RESOURCES: Record<string, Resource> = {
  organization_profile: resource(
    'Organization profile',
    'Defaults for the global organization record.',
    false,
    0,
    {
      countryCode: field('Country', 'reference', {
        referenceNamespace: 'country',
      }),
      language: field('Language', 'string'),
      timezone: field('Timezone', 'string'),
    }
  ),
}

export const APPLICATION_RESOURCES: Record<string, Record<string, Resource>> = {
  [BILLING_APP_SLUG]: {
    document_preference: resource(
      'Documents',
      'Default customer note and terms per document type. Subscription-generated invoices inherit the invoice preference.',
      true,
      0,
      {
        documentType: field('Document type', 'string', {
          allowedValues: ['invoice', 'quote', 'estimate', 'credit_note'],
          unique: true,
        }),
        customerNote: field('Customer note', 'string', { required: false }),
        termsAndConditions: field('Terms and conditions', 'string', {
          required: false,
        }),
      }
    ),
  },
  [CRM_APP_SLUG]: {
    request_priority: resource(
      'Request priorities',
      'Priorities created for newly provisioned CRM workspaces.',
      true,
      1,
      {
        name: field('Name', 'string'),
        description: field('Description', 'string', { required: false }),
        color: field('Color', 'string', { required: false }),
        icon: field('Icon', 'string', { required: false }),
        weight: field('Weight', 'integer'),
        sortOrder: field('Sort order', 'integer'),
        isDefault: field('Default', 'boolean'),
      }
    ),
    request_category: resource(
      'Request categories',
      'Categories created for newly provisioned CRM workspaces.',
      true,
      1,
      {
        name: field('Name', 'string'),
        description: field('Description', 'string', { required: false }),
        color: field('Color', 'string', { required: false }),
        icon: field('Icon', 'string', { required: false }),
        sortOrder: field('Sort order', 'integer'),
        isActive: field('Active', 'boolean'),
        defaultPriority: field('Default priority', 'reference', {
          required: false,
          referenceNamespace: 'request_priority',
        }),
      }
    ),
    request_subcategory: resource(
      'Request subcategories',
      'Optional subcategories created beneath provisioned CRM categories.',
      true,
      0,
      {
        category: field('Category', 'reference', {
          referenceNamespace: 'request_category',
        }),
        name: field('Name', 'string'),
        description: field('Description', 'string', { required: false }),
        icon: field('Icon', 'string', { required: false }),
        sortOrder: field('Sort order', 'integer'),
        isActive: field('Active', 'boolean'),
        defaultPriority: field('Default priority', 'reference', {
          required: false,
          referenceNamespace: 'request_priority',
        }),
      }
    ),
  },
}

export function resourceRegistry(
  targetType: ProvisioningTargetType,
  targetKey: string
): Record<string, Resource> {
  if (targetType === 'finance') return FINANCE_RESOURCES
  if (targetType === 'organization') return ORGANIZATION_RESOURCES
  return APPLICATION_RESOURCES[targetKey] ?? {}
}

export function catalogDefinitions(
  targetType: ProvisioningTargetType,
  targetKey: string
): ProvisioningResourceDefinition[] {
  const registry = resourceRegistry(targetType, targetKey)
  const definitions: ProvisioningResourceDefinition[] = []

  for (const [resourceType, resourceDef] of Object.entries(registry)) {
    definitions.push({
      resourceType,
      label: resourceDef.label,
      description: resourceDef.description,
      multiple: resourceDef.multiple,
      minimumItems: resourceDef.minimumItems,
      maximumItems: resourceDef.multiple ? null : 1,
      fields: Object.entries(resourceDef.fields).map(([key, fieldDef]) => ({
        key,
        label: fieldDef.label,
        valueType: fieldDef.valueType,
        required: fieldDef.required,
        referenceNamespace: fieldDef.referenceNamespace,
        allowedValues: fieldDef.allowedValues
          ? [...fieldDef.allowedValues]
          : null,
      })),
    })
  }

  return definitions
}

function stringProperty(
  resource: ProvisioningResourceInput,
  key: string
): string | null {
  const property = resource.properties.find((candidate) => candidate.key === key)
  return property?.valueType === 'string' ? property.stringValue : null
}

export function validateDraft(
  targetType: ProvisioningTargetType,
  targetKey: string,
  draft: ProvisioningDraftReplace
): ProvisioningValidationIssue[] {
  const issues: ProvisioningValidationIssue[] = []

  if (targetType !== 'application' && draft.financeDependency !== 'none') {
    issues.push({
      path: 'finance_dependency',
      code: 'invalid_target_contract',
      message:
        'Only application manifests may declare an embedded finance dependency.',
    })
  }

  const registry = resourceRegistry(targetType, targetKey)
  const counts: Record<string, number> = {}

  draft.resources.forEach((resource, index) => {
    const path = `resources.${index}`
    const definition = registry[resource.resourceType]

    if (!definition) {
      issues.push({
        path: `${path}.resource_type`,
        code: 'unknown_resource_type',
        message: `Resource type '${resource.resourceType}' is not registered for this target.`,
      })
      return
    }

    counts[resource.resourceType] = (counts[resource.resourceType] ?? 0) + 1

    const values = new Map(resource.properties.map((p) => [p.key, p]))

    for (const [key, fieldDef] of Object.entries(definition.fields)) {
      const prop = values.get(key)

      if (!prop) {
        if (fieldDef.required)
          issues.push({
            path: `${path}.properties.${key}`,
            code: 'missing_property',
            message: `Property '${key}' is required.`,
          })
        continue
      }

      if (prop.valueType !== fieldDef.valueType)
        issues.push({
          path: `${path}.properties.${key}.value_type`,
          code: 'invalid_property_type',
          message: `Property '${key}' must use value type '${fieldDef.valueType}'.`,
        })

      if (
        fieldDef.referenceNamespace &&
        prop.referenceNamespace !== fieldDef.referenceNamespace
      )
        issues.push({
          path: `${path}.properties.${key}.reference_namespace`,
          code: 'invalid_reference_namespace',
          message: `Property '${key}' must reference '${fieldDef.referenceNamespace}'.`,
        })

      const scalar = prop.stringValue
      if (
        fieldDef.allowedValues &&
        !fieldDef.allowedValues.includes(scalar as string)
      )
        issues.push({
          path: `${path}.properties.${key}`,
          code: 'invalid_property_value',
          message: `Property '${key}' must be one of: ${fieldDef.allowedValues.join(', ')}.`,
        })
    }

    for (const key of values.keys())
      if (!(key in definition.fields))
        issues.push({
          path: `${path}.properties.${key}`,
          code: 'unknown_property',
          message: `Property '${key}' is not registered for resource type '${resource.resourceType}'.`,
        })

    if (resource.resourceType === 'tax_rate') {
      const effectiveFrom = stringProperty(resource, 'effectiveFrom')
      const effectiveUntil = stringProperty(resource, 'effectiveUntil')
      const fromValid = !effectiveFrom || ISO_DATE_PATTERN.test(effectiveFrom)
      const untilValid = !effectiveUntil || ISO_DATE_PATTERN.test(effectiveUntil)

      if (!fromValid)
        issues.push({
          path: `${path}.properties.effectiveFrom`,
          code: 'invalid_date',
          message: 'Effective from must use YYYY-MM-DD.',
        })
      if (!untilValid)
        issues.push({
          path: `${path}.properties.effectiveUntil`,
          code: 'invalid_date',
          message: 'Effective until must use YYYY-MM-DD.',
        })
      if (
        effectiveFrom &&
        effectiveUntil &&
        fromValid &&
        untilValid &&
        effectiveFrom > effectiveUntil
      )
        issues.push({
          path: `${path}.properties.effectiveUntil`,
          code: 'invalid_date_range',
          message: 'Effective until cannot be earlier than effective from.',
        })
    }
  })

  for (const [resourceType, count] of Object.entries(counts))
    if (count > 1 && !registry[resourceType]!.multiple)
      issues.push({
        path: 'resources',
        code: 'resource_cardinality',
        message: `Resource type '${resourceType}' permits only one row.`,
      })

  for (const [resourceType, definition] of Object.entries(registry)) {
    const count = counts[resourceType] ?? 0
    if (count < definition.minimumItems)
      issues.push({
        path: 'resources',
        code: 'resource_minimum',
        message: `Resource type '${resourceType}' requires at least ${definition.minimumItems} row(s).`,
      })
  }

  const referenceNamespaces = new Set<string>()
  for (const definition of Object.values(registry))
    for (const fieldDef of Object.values(definition.fields))
      if (
        fieldDef.referenceNamespace &&
        fieldDef.referenceNamespace in registry
      )
        referenceNamespaces.add(fieldDef.referenceNamespace)

  const referenceTargets: Record<string, Set<string>> = {}
  for (const ns of referenceNamespaces)
    referenceTargets[ns] = new Set(
      draft.resources.filter((r) => r.resourceType === ns).map((r) => r.key)
    )

  draft.resources.forEach((resource, index) => {
    for (const prop of resource.properties) {
      const ns = prop.referenceNamespace
      if (ns && ns in referenceTargets) {
        const targets = referenceTargets[ns]!
        if (!targets.has(prop.referenceKey as string))
          issues.push({
            path: `resources.${index}.properties.${prop.key}`,
            code: 'unresolved_reference',
            message: `Reference '${ns}/${prop.referenceKey}' does not exist in this draft.`,
          })
      }
    }
  })

  for (const [resourceType, definition] of Object.entries(registry)) {
    for (const [fieldKey, fieldDef] of Object.entries(definition.fields)) {
      if (!fieldDef.unique) continue
      const seen = new Set<string>()

      draft.resources.forEach((resource, index) => {
        if (resource.resourceType !== resourceType) return
        const prop = resource.properties.find((p) => p.key === fieldKey)
        if (!prop) return

        const tuple = JSON.stringify([
          prop.valueType,
          prop.stringValue,
          prop.integerValue,
          prop.decimalValue,
          prop.booleanValue,
          prop.referenceNamespace,
          prop.referenceKey,
        ])

        if (seen.has(tuple))
          issues.push({
            path: `resources.${index}.properties.${fieldKey}`,
            code: 'duplicate_unique_property',
            message: `Property '${fieldKey}' must be unique within resource type '${resourceType}'.`,
          })
        else seen.add(tuple)
      })
    }
  }

  if (targetType === 'application' && targetKey === CRM_APP_SLUG) {
    const defaults = draft.resources.filter((resource) => {
      if (resource.resourceType !== 'request_priority') return false
      return resource.properties.some(
        (property) =>
          property.key === 'isDefault' && property.booleanValue === true
      )
    })
    if (defaults.length !== 1)
      issues.push({
        path: 'resources',
        code: 'crm_default_priority',
        message:
          'CRM provisioning requires exactly one default request priority.',
      })
  }

  return issues
}

export function validateProvisioningWireDraft(
  targetType: ProvisioningTargetType,
  targetKey: string,
  draft: ProvisioningWireDraftReplace
): ProvisioningValidationIssue[] {
  return validateDraft(targetType, targetKey, provisioningDraftForCatalog(draft))
}
